File: algorithm/two-sum/two-sum.cpp
class Solution {
    
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> ans;
        for (vector<int>::iterator it = nums.begin(); it !=  nums.end(); it++) {
            vector<int>::iterator jt = find(it + 1, nums.end(), target - *it);
            if (jt != nums.end()) {
                ans.push_back(it - nums.begin());
                ans.push_back(jt - nums.begin());
                return ans;
            }
        }
        return ans;
    }
};

File: algorithm/two-sum/two-sum.go
package two_sum

func twoSum(nums []int, target int) []int {
	cache := make(map[int][]int)

	for i, n := range nums {
		cache[n] = append(cache[n], i)
	}

	for i, n := range nums {
		indices, ok := cache[target-n]
		if !ok {
			continue
		}

		if len(indices) == 2 {
			return indices
		}

		if indices[0] == i {
			continue
		}

		return []int{i, indices[0]}
	}

	return []int{-1, -1}
}
